src/worker_thread.c

```diff
@@ -0,0 +1,289 @@
+{
+    WORKER_THREAD_HANDLE worker_thread;
+
+    /* Codes_SRS_WORKER_THREAD_01_004: [ worker_func_context shall be allowed to be NULL. ]*/
+
+    /* Codes_SRS_WORKER_THREAD_01_003: [ If worker_func is NULL, worker_thread_create shall fail and return NULL. ]*/
+    if (worker_func == NULL)
+    {
+        LogError("NULL worker_func");
+    }
+    else
+    {
+        /* Codes_SRS_WORKER_THREAD_01_001: [ worker_thread_create shall allocate memory for a new worker thread object and on success return a non-NULL handle to it. ]*/
+        worker_thread = malloc(sizeof(WORKER_THREAD));
+        if (worker_thread == NULL)
+        {
+            /* Codes_SRS_WORKER_THREAD_01_008: [ If any error occurs, worker_thread_create shall fail and return NULL. ]*/
+            LogError("Cannot allocate memory for BS worker thread");
+        }
+        else
+        {
+            worker_thread->worker_func = worker_func;
+            worker_thread->worker_func_context = worker_func_context;
+
+            /* Codes_SRS_WORKER_THREAD_01_037: [ worker_thread_create shall create a state manager object by calling sm_create with the name worker_thread. ]*/
+            worker_thread->sm = sm_create("worker_thread");
+            if (worker_thread->sm == NULL)
+            {
+                LogError("sm_create failed");
+            }
+            else
+            {
+                /* Codes_SRS_WORKER_THREAD_01_022: [ worker_thread_create shall perform the following actions in order: ]*/
+
+                /* Codes_SRS_WORKER_THREAD_01_006: [ worker_thread_create shall create an auto reset unnamed event used for signaling the thread to shutdown by calling CreateEvent. ]*/
+                worker_thread->worker_thread_shutdown_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+                if (worker_thread->worker_thread_shutdown_event == NULL)
+                {
+                    /* Codes_SRS_WORKER_THREAD_01_008: [ If any error occurs, worker_thread_create shall fail and return NULL. ]*/
+                    LogError("Cannot create shutdown event");
+                }
+                else
+                {
+                    /* Codes_SRS_WORKER_THREAD_01_007: [ worker_thread_create shall create an auto reset unnamed event used for signaling that a new execution of the worker function should be done by calling CreateEvent. ]*/
+                    worker_thread->execute_worker_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+                    if (worker_thread->execute_worker_event == NULL)
+                    {
+                        /* Codes_SRS_WORKER_THREAD_01_008: [ If any error occurs, worker_thread_create shall fail and return NULL. ]*/
+                        LogError("Cannot create process event");
+                    }
+                    else
+                    {
+                        goto all_ok;
+
+                        //(void)CloseHandle(worker_thread->execute_worker_event);
```

Comment:
>//(void)CloseHandle(worker_thread->execute_worker_event); [](http://example.com/codeflow?start=24&length=57)

Can you please remove the comment

Reply:
That one is left on purpose. This is the only exception we have for leaving comments in the code (it is cleanup that would be enabled at the moment you add another if in the function).